Gems/EMotionFX/Code/EMotionFX/Pipeline/RCExt/Actor/ActorBuilder.cpp

```diff
@@ -206,37 +206,42 @@ namespace EMotionFX
             if (skeletonOptimizationRule)
             {
                 const SceneData::SceneNodeSelectionList& criticalBonesList = skeletonOptimizationRule->GetCriticalBonesList();
-                const size_t criticalBoneCount = criticalBonesList.GetSelectedNodeCount();
-                
-                for (size_t boneIndex = 0; boneIndex < criticalBoneCount; ++boneIndex)
-                {
-                    const AZStd::string& bonePath = criticalBonesList.GetSelectedNode(boneIndex);
 
-                    // The bone name stores the node path in the scene. We need to convert it to EMotionFX node name.
-                    const SceneContainers::SceneGraph::NodeIndex nodeIndex = graph.Find(bonePath);
-                    if (!nodeIndex.IsValid())
+                criticalBonesList.EnumerateSelectedNodes(
+                    [&](const AZStd::string& bonePath)
                     {
-                        AZ_TracePrintf(SceneUtil::WarningWindow, "Critical bone %s is not stored in the scene. Skipping it.", bonePath.c_str());
-                        continue;
-                    }
+                        // The bone name stores the node path in the scene. We need to convert it to EMotionFX node name.
+                        const SceneContainers::SceneGraph::NodeIndex nodeIndex = graph.Find(bonePath);
+                        if (!nodeIndex.IsValid())
+                        {
+                            AZ_TracePrintf(
+                                SceneUtil::WarningWindow, "Critical bone %s is not stored in the scene. Skipping it.", bonePath.c_str());
+                            return true;
+                        }
 
-                    AZStd::shared_ptr<const SceneDataTypes::IBoneData> boneData = azrtti_cast<const SceneDataTypes::IBoneData*>(graph.GetNodeContent(nodeIndex));
-                    if (!boneData)
-                    {
-                        // Make sure we are dealing with bone here.
-                        continue;
-                    }
+                        AZStd::shared_ptr<const SceneDataTypes::IBoneData> boneData =
+                            azrtti_cast<const SceneDataTypes::IBoneData*>(graph.GetNodeContent(nodeIndex));
+                        if (!boneData)
+                        {
+                            // Make sure we are dealing with bone here.
+                            return true;
+                        }
```

Comment:
This code _smells_
It stores a raw pointer inside of the boneData shared_ptr, but only checks if the pointer is nullptr.
There is no reason to construct a shared_ptr here.
```suggestion             
                        if (auto boneData =
                            azrtti_cast<const SceneDataTypes::IBoneData*>(graph.GetNodeContent(nodeIndex)); boneData == nullptr)
                        {
                            // Make sure we are dealing with bone here.
                            return true;
                        }
```

Reply:
Actually it looks like `SceneGraph::GetNodeContent` returns a shared_ptr, so this cast will fail.
[AZStd::rtti_pointer_cast](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/smart_ptr/shared_ptr.h#L434-L436) should be used instead
```suggestion
                        AZStd::shared_ptr<const SceneDataTypes::IBoneData> boneData =
                            AZStd::rtti_pointer_cast<const SceneDataTypes::IBoneData>(graph.GetNodeContent(nodeIndex));
                        if (!boneData)
                        {
                            // Make sure we are dealing with bone here.
                            return true;
                        }
```

Reply:
Nevermind, it looks like we have a specialization for RttiCast for the `AZStd::shared_ptr` class template in [RTTI.h](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/RTTI/RTTI.h#L645-L660)

Reply:
Hrm, I'm a little nervous to bugfix this code, I don't know the EMFX code well enough to know how to test this to see what happens when this logic suddenly starts working correctly. :(

Reply:
So should I just change this to code to this and skip the boneData temp var entirely?
```
                        if (!azrtti_cast<const SceneDataTypes::IBoneData*>(graph.GetNodeContent(nodeIndex)))
                        {
                            // Make sure we are dealing with bone here.
                            return true;
                        }
```

Reply:
No, you just can leave the code as is.
It should be working as is.